src/algebra/regular.lean

```diff
@@ -0,0 +1,186 @@
+/-
+Copyright (c) 2021 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import algebra.group
+
+/-!
+# Regular elements
+
+We introduce left-regular, right-regular and regular elements.
+
+By definition, a regular element in a commutative ring is a non-zero divisor.
+Lemma `is_regular_of_integral_domain` shows that every non-zero element of an integral domain
+is regular.
+
+The final goal is to develop part of the API to prove, eventually, results about non-zero-divisors.
+-/
+variables {R : Type*}
+
+/-- A left-regular element is an element `c` such that multiplication on the left by `c`
+is injective on the left. -/
+def is_left_regular [has_mul R] (c : R) := function.injective ((*) c)
+
+/-- A right-regular element is an element `c` such that multiplication on the right by `c`
+is injective on the right. -/
+def is_right_regular [has_mul R] (c : R) := function.injective (* c)
+
+/-- A regular element is an element `c` such that multiplication by `c` both on the left and
+on the right is injective. -/
+structure is_regular [has_mul R] (c : R) : Prop :=
+(left : is_left_regular c)
+(right : is_right_regular c)
+
+namespace is_regular
+
+section semigroup
+
+variable [semigroup R]
+
+/-- In a semigroup, then the product of left-regular elements is left-regular. -/
+lemma is_left_regular.mul {a b : R}
+  (lra : is_left_regular a) (lrb : is_left_regular b) :
+  is_left_regular (a * b) :=
+λ c d cd, lrb (lra ((mul_assoc a b c).symm.trans (cd.trans (mul_assoc a b d))))
+
+/-- In a semigroup, then the product of right-regular elements is right-regular. -/
+lemma is_right_regular.mul {a b : R}
+  (rra : is_right_regular a) (rrb : is_right_regular b) :
+  is_right_regular (a * b) :=
+λ c d cd, rra (rrb ((mul_assoc c a b).trans (cd.trans (mul_assoc d a b).symm)))
+
+/--  If an element `b` becomes left-regular after multiplying it on the left by a left-regular
+element, then `b` is left-regular. -/
+lemma is_left_regular_saturated {a b : R} (ab : is_left_regular (a * b)) :
+  is_left_regular b :=
+function.injective.of_comp (by rwa comp_mul_left a b)
+
+/--  An element is left-regular if and only if multiplying it on the left by a left-regular element
+is left-regular. -/
+@[simp] lemma mul_is_left_regular_iff {a : R} (b : R) (ha : is_left_regular a) :
+  is_left_regular (a * b) ↔ is_left_regular b :=
+⟨λ ab, is_left_regular_saturated ab, λ ab, is_left_regular.mul ha ab⟩
```

Comment:
This lemma is very similar to the one I propose in #6283!

Reply:
Cool!  I wonder whether we should (at a later stage, maybe) thinking about making both a special case of a more general result!